### ISSUE TITLE
Broken node affinity

### ISSUE THREAD

**Issue by RenaudWasTaken** [kubernetes/kubernetes#44349]
**Is this a BUG REPORT or FEATURE REQUEST?**: Bug report

**What happened**:
TLDR: Node affinity is broken for more than one requirement

- git clone https://github.com/kubernetes/kubernetes.git
- cd ./kubernetes && ./hack/local-up-cluster.sh
- ./cluster/kubectl.sh label nodes 127.0.0.1 foo=5
- ./cluster/kubectl.sh label nodes 127.0.0.1 bar=6
- cat ./pod.yml
```yml
apiVersion: v1
kind: Pod
metadata:
  name: nginx-pod
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: foo
            operator: Gt
            values:
            - "4"
        - matchExpressions:
          - key: bar
            operator: Gt
            values:
            - "8"

  containers:
  - image: nginx
    name: nginx-ctr
```
5) ./cluster/kubectl.sh create -f ./pod.yml
6) ./cluster/kubectl.sh get nodes
```bash
NAME        READY     STATUS    RESTARTS   AGE
nginx-pod   1/1       Running   0          11s
```
**What you expected to happen**:
The pod should not have been scheduled

**Anything else we need to know**:
This looks like a logic bug in the scheduler: plugin/pkg/scheduler/algorithm/predicates/predicates.go:596

The function in cause is: `nodeMatchesNodeSelectorTerms` called from `podMatchesNodeLabels`
I think it should return false when one of the predicates does not match on the node and return true at the end

**Environment and version**:

```bash
$ cluster/kubectl.sh version
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.0.1479+03aba86974957f-dirty", GitCommit:"03aba86974957feae11a73a0db43a39b5743c550", GitTreeState:"dirty", BuildDate:"2017-03-23T21:54:07Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}

$ cat /etc/os-release                                                                                                                                        NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
UBUNTU_CODENAME=xenial

$ uname -a                                                                                                                                                   
Linux 4.4.0-62-generic #83-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux
```
Ran from `./hack/local-up-cluster.sh`

**Comment by aveshagarwal** [comment]
@kubernetes/sig-scheduling-bugs

**Comment by jayunit100** [comment]
cc @ravisantoshgudimetla  I've reproduced this and it will effect your current scheduler_perf work.  Can you tackle it ? 

Culprit might be around here in predicates.go .

```
		// Match node selector for requiredDuringSchedulingIgnoredDuringExecution.
		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
			nodeSelectorTerms := nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
			glog.V(10).Infof("Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms %+v", nodeSelectorTerms)
			nodeAffinityMatches = nodeAffinityMatches && nodeMatchesNodeSelectorTerms(node, nodeSelectorTerms)
		}
```

By the way note also need to udpate docs for `RequiredDuringSchedulingRequiredDuringExecution`, as its not supported but iirc in the .md files it provides it as an option.  I was thinking it might be a good positive control to debug things with.

**Comment by ravisantoshgudimetla** [comment]
@jayunit100 Sure. I will work on this.

**Comment by davidopp** [comment]
I don't think this is a bug. The API definition says `nodeSelectorTerm`'s are ORed

```
// A node selector represents the union of the results of one or more label queries
// over a set of nodes; that is, it represents the OR of the selectors represented
// by the node selector terms.
type NodeSelector struct {
	//Required. A list of node selector terms. The terms are ORed.
	NodeSelectorTerms []NodeSelectorTerm
}

// A null or empty node selector term matches no objects.
type NodeSelectorTerm struct {
	//Required. A list of node selector requirements. The requirements are ANDed.
	MatchExpressions []NodeSelectorRequirement
}
```

**Comment by aveshagarwal** [comment]
I think what you are trying to do, you need:
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-pod
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: foo
            operator: Gt
            values:
            - "4"
          - key: bar
            operator: Gt
            values:
            - "8"
  containers:
  - image: nginx
    name: nginx-ctr
```

**Comment by gyliu513** [comment]
I think that we should update the document here to mention how does `multiple nodeSelectorTerms` and `multiple nodeSelector` works for such case.

**Comment by davidopp** [comment]
Agreed. @aveshagarwal also brought it up on the mailing list today:
https://groups.google.com/d/msg/kubernetes-sig-scheduling/P249hlfGcdw/_wfQ7UOlBQAJ

I'm happy to review PRs to improve the documentation.

**Comment by gyliu513** [comment]
Yes, just saw that email, will try to update the document for what @aveshagarwal point out, this will make the document more clear for node affinity features. @davidopp

**Comment by gyliu513** [comment]
I think this one can be closed due to it is a document issue. @davidopp @RenaudWasTaken

**Comment by jayunit100** [comment]
Ah Yeah, confirmed its an `OR` vs `AND` issue, @aveshagarwal 's snippet "works", i.e. blocks scheduling properly.  thanks for clarifying.

**Comment by RenaudWasTaken** [comment]
Right !
Thanks for clarifying this @davidopp :)
